{"cells":[{"cell_type":"markdown","metadata":{"id":"vPeecVCfKURM"},"source":["# Notebook for experiment tracking\n","## Store remote only artifacts (does not work)\n","MLflow original setup:\n","* Tracking server: local filesystem.\n","* Backend store: local sqlite database.\n","* Artifacts store: s3 bucket.\n","\n","The Reviewers must be able to see the experiments so the MLflow Tracking Server (creates and manages experiments and runs) will be local. Having it remotely it would mean having an EC2 instance shared with other users which could incurr in unnecessary costs. \n","\n","For the same reason the backend will be stored locally.\n","\n","Since we want the models to be available in the cloud the Artifacts store will be a s3 bucket.\n","\n","NOTE: I follow video 4.3: https://app.clickup.com/t/2k55fz6 and 04-deployment/web-service-mlflow/random-forest_with_pipeline.ipynb\n","\n","To run this you need to launch the mlflow server locally by running the following command in your terminal:\n","\n","`mlflow server --backend-store-uri=sqlite:///mlflow_db.db --default-artifact-root=s3://mlflow-artifacts-remote-jaime/`\n"]},{"cell_type":"markdown","metadata":{},"source":["--- The first parts of the notebook taken from 08-my_project/EDA/Version_1.ipynb ---"]},{"cell_type":"markdown","metadata":{},"source":["## Import Libs"]},{"cell_type":"code","execution_count":26,"metadata":{},"outputs":[],"source":["from datetime import timedelta, datetime, timezone\n","from pandas._libs.tslibs.timestamps import Timestamp\n","import pandas as pd\n","import numpy as np\n","import mlflow\n","from sklearn.linear_model import LinearRegression\n","from sklearn.metrics import mean_squared_error\n","from sklearn.preprocessing import PolynomialFeatures\n","from sklearn.pipeline import Pipeline"]},{"cell_type":"markdown","metadata":{},"source":["## Get and prepare data"]},{"cell_type":"code","execution_count":27,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","/tmp/ipykernel_4074/630401041.py:15: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n"]}],"source":["LAST_DAYS = 30 #Number of days to get data from\n","\n","now = datetime.now()\n","\n","dfs = []  # empty list which will hold your dataframes\n","df_temp_2 = pd.DataFrame()\n","for d in range(1, LAST_DAYS): #NOTE: do the same that has been done for the first day but for the whole period\n","    date = now - timedelta(days=d)\n","    date_str = date.strftime(\"%m-%d-%Y\")\n","    # print(date_str)\n","    source_url = 'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_daily_reports/' + date_str + '.csv'\n","    df_temp = pd.read_csv(source_url)\n","    df_temp.rename(columns={\"Last_Update\": \"Date\"}, inplace=True) #Renane dataframe column from \"Last_Update\" to \"Date\"\n","    df_temp_2 = df_temp[[\"Admin2\", \"Province_State\", \"Country_Region\",\"Confirmed\", \"Deaths\"]] #TODO: consider also other columns in future versions like Recovered,Active,Combined_Key,Incident_Rate,Case_Fatality_Ratio\n","    df_temp_2[\"Date\"] = date.strftime(\"%Y-%m-%d\") #TODO: fix this so that no warning comes\n","    dfs.append(df_temp_2)  # append dataframe to list\n","    \n","res = pd.concat(dfs, ignore_index=True)  # concatenate list of dataframes\n"]},{"cell_type":"code","execution_count":28,"metadata":{},"outputs":[],"source":["# group by Country_Region and sum Confirmed and Deaths\n","df = res.groupby(['Province_State','Country_Region','Date']).agg({'Confirmed':'sum', 'Deaths':'sum'})\n","df.reset_index(inplace=True)\n","df.rename(columns={\"Confirmed\": \"ConfirmedCases\", \"Deaths\": \"Fatalities\"}, inplace=True)"]},{"cell_type":"code","execution_count":29,"metadata":{},"outputs":[],"source":["## Prepare train(dev)-test set\n","loc_group = [\"Province_State\", \"Country_Region\"]\n","def preprocess(df):\n","    df[\"Date\"] = df[\"Date\"].astype(\"datetime64[ms]\")\n","    for col in loc_group:\n","        df[col].fillna(\"none\", inplace=True) #NOTE: replace all NaN with none\n","    return df\n","df = preprocess(df)\n","\n","TARGETS = [\"ConfirmedCases\", \"Fatalities\"]\n","for col in TARGETS:\n","    df[col] = np.log1p(df[col]) \n","    \n","for col in TARGETS:\n","    df[\"prev_{}\".format(col)] = df.groupby(loc_group)[col].shift() #NOTE: the prev_ columns basically has the same than the others but delayed one day\n","    \n","    \n","df = df[df[\"Date\"] > df[\"Date\"].min()].copy() #NOTE: removes the first day since it has NaNs in the \"prev\" columns\n","\n","TEST_DAYS = 7 #Number of days to test the model\n","TEST_FIRST = now - timedelta(days=TEST_DAYS)\n","TEST_FIRST = TEST_FIRST.replace(hour=0, minute=0, second=0, microsecond=0)\n","TEST_FIRST = Timestamp(TEST_FIRST)\n","\n","dev_df, test_df = df[df[\"Date\"] < TEST_FIRST].copy(), df[df[\"Date\"] >= TEST_FIRST].copy() #I am testing the model with the predictions of the last 7 (TEST_DAYS) days and\n","                                                                                          # training it with data from the previous 30 days (LAST_DAYS) excluding the last 7 days (TEST_DAYS)\n"]},{"cell_type":"code","execution_count":30,"metadata":{},"outputs":[],"source":["features = [\"prev_{}\".format(col) for col in TARGETS]"]},{"cell_type":"markdown","metadata":{},"source":["## Modeling"]},{"cell_type":"code","execution_count":31,"metadata":{},"outputs":[],"source":["mlflow.set_tracking_uri(\"http://127.0.0.1:5000\") #NOTE: Important!!!  to set the tracking uri here cause otherwise it stores the artifact locally"]},{"cell_type":"code","execution_count":32,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["tracking URI: 'http://127.0.0.1:5000'\n"]}],"source":["#NOTE: since we did not specify a tracking uri by default mlflow will store everything in the local system\n","print(f\"tracking URI: '{mlflow.get_tracking_uri()}'\")"]},{"cell_type":"code","execution_count":33,"metadata":{},"outputs":[],"source":["#TODO: fix the UserWarnings that appear when running this cell\n","\n","def rmse(y_true, y_pred):\n","    return np.sqrt(mean_squared_error(y_true, y_pred))\n","\n","def evaluate(df):\n","    error = 0\n","    for col in TARGETS:\n","        error += rmse(df[col].values, df[\"pred_{}\".format(col)].values) #NOTE: checks the error between the predicted columns and the target columns\n","    return np.round(error/len(TARGETS), 5)\n","\n","\n","def predict(test_df, first_day, num_days, val=False):\n","    y_pred = np.clip(model.predict(test_df.loc[test_df[\"Date\"] == first_day][features]), None, 16)#NOTE: here he is predicting the targets for the first day and \n","                                                                                                    #saturating (clip) them with max=16\n"," \n","    for i, col in enumerate(TARGETS):\n","        test_df[\"pred_{}\".format(col)] = 0\n","        test_df.loc[test_df[\"Date\"] == first_day, \"pred_{}\".format(col)] = y_pred[:, i] #NOTE: here he sets the predicted columns\n","\n","    if val:\n","        print(first_day, evaluate(test_df[test_df[\"Date\"] == first_day])) #NOTE: print the date of the first day and the error between the predicted targets and the real targets\n","\n","    for d in range(1, num_days): #NOTE: do the same that has been done for the first day but for the whole period\n","        y_pred = np.clip(model.predict(y_pred), None, 16)\n","        date = first_day + timedelta(days=d)\n","\n","        for i, col in enumerate(TARGETS):\n","            test_df.loc[test_df[\"Date\"] == date, \"pred_{}\".format(col)] = y_pred[:, i]\n","\n","        if val:\n","            print(date, evaluate(test_df[test_df[\"Date\"] == date])) #NOTE: when we see all the errors we can see that the farther the date from the first day the higher the error\n","        \n","    return test_df"]},{"cell_type":"code","execution_count":34,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["2022/08/10 12:46:33 INFO mlflow.tracking.fluent: Experiment with name 'experiment-covid-2' does not exist. Creating a new experiment.\n","/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n","/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n","/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n","/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n","/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n","/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n"]},{"name":"stdout","output_type":"stream","text":["2022-08-03 00:00:00 0.02766\n","2022-08-04 00:00:00 0.04321\n","2022-08-05 00:00:00 0.0611\n","2022-08-06 00:00:00 0.07638\n","2022-08-07 00:00:00 0.09059\n","2022-08-08 00:00:00 0.10535\n","2022-08-09 00:00:00 0.30238\n","RMSE: 0.13264\n","default artifacts URI: 's3://mlflow-artifacts-remote-jaime/4/16082a31f2be4eadb6f368b4ded2d309/artifacts'\n"]}],"source":["mlflow.set_experiment(\"experiment-covid-2\") \n","\n","with mlflow.start_run():\n","    \n","    degree_poly = 2\n","    incl_bias = False\n","    mlflow.log_param('degree_poly', degree_poly)\n","    mlflow.log_param('include_bias', incl_bias)\n","    \n","    model = Pipeline([('poly', PolynomialFeatures(degree=degree_poly, include_bias=incl_bias)),\n","                  ('linear', LinearRegression())])\n","    \n","    model.fit(dev_df[features], dev_df[TARGETS])\n","    [mean_squared_error(dev_df[TARGETS[i]], model.predict(dev_df[features])[:, i]) for i in range(len(TARGETS))] #NOTE: check the mean_squared_error from the training dataset\n","    \n","    test_df = predict(test_df, TEST_FIRST, TEST_DAYS, val=True) #NOTE: makes predictions \n","    #for TEST_DAYS number of days ...just to print it on the screen\n","    \n","    eval_RMSE = evaluate(test_df) #NOTE: the error of all the predictions\n","    print(\"RMSE:\", eval_RMSE)\n","    mlflow.log_metric(\"evaluated_RMSE\", eval_RMSE)\n","\n","    mlflow.sklearn.log_model(model, artifact_path=\"models\")\n","    print(f\"default artifacts URI: '{mlflow.get_artifact_uri()}'\") #returns where the model is stored\n","    "]},{"cell_type":"code","execution_count":35,"metadata":{},"outputs":[{"data":{"text/plain":["[<Experiment: artifact_location='s3://mlflow-artifacts-remote-jaime/0', experiment_id='0', lifecycle_stage='active', name='Default', tags={}>,\n"," <Experiment: artifact_location='s3://mlflow-artifacts-remote-jaime/1', experiment_id='1', lifecycle_stage='active', name='my-experiment-1', tags={}>,\n"," <Experiment: artifact_location='s3://mlflow-artifacts-remote-jaime/2', experiment_id='2', lifecycle_stage='active', name='experiment-covid-1', tags={}>,\n"," <Experiment: artifact_location='s3://mlflow-artifacts-remote-jaime/3', experiment_id='3', lifecycle_stage='active', name='my-cool-experiment', tags={}>,\n"," <Experiment: artifact_location='s3://mlflow-artifacts-remote-jaime/4', experiment_id='4', lifecycle_stage='active', name='experiment-covid-2', tags={}>]"]},"execution_count":35,"metadata":{},"output_type":"execute_result"}],"source":["mlflow.list_experiments() \n","#NOTE: this list does not represent the structure in the S3 bucket itself. You have to get the RUNID from mlflow ui and then look for that id under the \n","# folders in the s3 bucket https://s3.console.aws.amazon.com/s3/buckets/mlflow-artifacts-remote-jaime?region=eu-central-1&tab=objects\n"]},{"cell_type":"markdown","metadata":{},"source":["I see in mlflow ui that the RUNID for this experiment is 16082a31f2be4eadb6f368b4ded2d309"]},{"cell_type":"code","execution_count":36,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["2022-08-03 00:00:00 0.0381\n","2022-08-04 00:00:00 0.06246\n","2022-08-05 00:00:00 0.08495\n","2022-08-06 00:00:00 0.10244\n","2022-08-07 00:00:00 0.11765\n","2022-08-08 00:00:00 0.13222\n","2022-08-09 00:00:00 0.31463\n","RMSE: 0.14873\n"]},{"name":"stderr","output_type":"stream","text":["/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n","/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n","/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n","/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n","/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n","/home/ubuntu/.local/share/virtualenvs/08-my_project-6zcrQ2k9/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names\n","  warnings.warn(\n"]},{"name":"stdout","output_type":"stream","text":["default artifacts URI: 's3://mlflow-artifacts-remote-jaime/4/ae2e389613094cf48d62eed43ce7850e/artifacts'\n"]}],"source":["#NOTE: here UI change the degree_poly to have another experiment\n","mlflow.set_experiment(\"experiment-covid-2\") \n","\n","with mlflow.start_run(): \n","    \n","    degree_poly = 3\n","    incl_bias = False\n","    mlflow.log_param('degree_poly', degree_poly)\n","    mlflow.log_param('include_bias', incl_bias)\n","    \n","    model = Pipeline([('poly', PolynomialFeatures(degree=degree_poly, include_bias=incl_bias)),\n","                  ('linear', LinearRegression())])\n","    \n","    model.fit(dev_df[features], dev_df[TARGETS])\n","    [mean_squared_error(dev_df[TARGETS[i]], model.predict(dev_df[features])[:, i]) for i in range(len(TARGETS))] #NOTE: check the mean_squared_error from the training dataset\n","    \n","    test_df = predict(test_df, TEST_FIRST, TEST_DAYS, val=True) #NOTE: makes predictions \n","    #for TEST_DAYS number of days ...just to print it on the screen\n","    \n","    eval_RMSE = evaluate(test_df) #NOTE: the error of all the predictions\n","    print(\"RMSE:\", eval_RMSE)\n","    mlflow.log_metric(\"evaluated_RMSE\", eval_RMSE)\n","\n","    mlflow.sklearn.log_model(model, artifact_path=\"models\")\n","    print(f\"default artifacts URI: '{mlflow.get_artifact_uri()}'\") #returns where the model is stored\n","    "]},{"cell_type":"markdown","metadata":{},"source":["I see in mlflow ui that the RUNID for this experiment is ae2e389613094cf48d62eed43ce7850e"]},{"cell_type":"markdown","metadata":{},"source":["Comparing the experiments in mlflow the first one has less evaluated RSME so I will choose that model => RUNID = 16082a31f2be4eadb6f368b4ded2d309\n","This will also be more in detail in 08-my_project/exp-track-mod-reg/model-registry.ipynb"]}],"metadata":{"colab":{"collapsed_sections":["lx0eEn2mDsYw"],"name":"my_Titanic.ipynb","provenance":[]},"kernelspec":{"display_name":"Python 3.10.4 ('08-my_project-6zcrQ2k9')","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.4"},"vscode":{"interpreter":{"hash":"7fde853788301886b8fed61f068d574e1d63c7368ed08973156a22dfdab1a95e"}}},"nbformat":4,"nbformat_minor":0}
